src/bt.rs

```diff
@@ -74,324 +73,217 @@ pub enum BinaryTreeError<V> {
+pub struct BinaryTree<V> {
+    nodes: Vec<Node<V>>,
+    root: Option<usize>,
+}
+
+impl<V> BinaryTree<V> {
+    /// Number of nodes pre-allocated each time the tree grows.
+    const NODES_CAPACITY: usize = 256;
 
     /// Inserts the value at the end of the path.
     ///
-    /// This function traverses the tree from this node (`self`) until reaching the
+    /// This function traverses the tree from the root node until reaching the
     /// node at the end of the path. If the node is unreachable or already
     /// contains a value, an error wrapping the value is returned. Otherwise,
     /// a new node containing the value is inserted.
     ///
     /// # Returns
     /// - `Ok(())` when the node is inserted at the end of the path.
-    /// - `Err(InsertNonEmptyNode(value))` when the subtree already contains a
+    /// - `Err(InsertNonEmptyNode(value))` when the tree already contains a
     /// value at the end of the path.
     /// - `Err(UnreachablePath(value))` when the end of the path is unreachable.
     ///
     /// In the error cases, no insertion occurs and the value is returned to
     /// the caller of this function.
     pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
-        enum Ref<'a, V> {
-            This(&'a mut Node<V>),
-            Other(&'a mut Option<Box<Node<V>>>),
-        }
-
-        // Finds the node at the end of the path.
-        let mut node = Ref::This(self);
+        let last = self.nodes.len();
+        let mut node = &mut self.root;
         for bit in path.iter() {
-            node = match node {
-                Ref::This(n) => Ref::Other(if !bit { &mut n.left } else { &mut n.right }),
-                Ref::Other(Some(n)) => Ref::Other(if !bit { &mut n.left } else { &mut n.right }),
-                Ref::Other(None) => {
-                    return Err(BinaryTreeError::UnreachableNode(value));
+            match *node {
+                None => return Err(BinaryTreeError::UnreachableNode(value)),
+                Some(next) => {
+                    let n = &mut self.nodes[next];
+                    node = if !bit { &mut n.left } else { &mut n.right };
                 }
-            };
+            }
         }
 
-        // Checks whether the node already has a value,
-        // If so, returns an error wrapping the value that could not be inserted.
-        // otherwise, a new node is created containing the value to be inserted.
-        match node {
-            Ref::This(_) | Ref::Other(Some(_)) => {
-                return Err(BinaryTreeError::InsertNonEmptyNode(value))
-            }
-            Ref::Other(empty_node) => {
-                *empty_node = Some(Box::new(Node::new(value)));
+        if node.is_some() {
+            return Err(BinaryTreeError::InsertNonEmptyNode(value));
+        } else {
+            *node = Some(last);
+
+            if self.nodes.len() == self.nodes.capacity() {
+                self.nodes.reserve(Self::NODES_CAPACITY);
```

Comment:
`Vec::push` already grows the capacity as needed, using an amortized strategy ensuring exponential growth of capacity. `Vec::reserve` falls through to the same codepath when it needs to grow the vector; while this reservation logic requests `NODES_CAPACITY = 256` rather than 1 more element each time, I think `Vec`'s [amortized-growth logic](https://doc.rust-lang.org/src/alloc/raw_vec.rs.html#464) means that we'll be growing the vector's underlying buffer by the same amount as if we just pushed each element without this reservation logic.

I think the logic to manually grow the vec by `NODES_CAPACITY` should be dropped, and we should just `push` each new element on, trusting the `Vec` implementation to provide appropriate performance/space tradeoff when growing the vector.

Reply:
ok, let's delegate the growth logic to rust Vec implementation